File: libCheck.js

"use strict"

// import secrets
// import string
// from lib import *
// import time
// import stat
// import uuid


  // Google for "ansi vt100 codes" to learn about the codes in the printf-strings
const charEsc='\u001b'
//const charEsc=`\033` // Legacy octal escape sequence.
//const charEsc=`\x1b` // This would work too.
//const charEsc=`\u{1b}` // This would also work.
//const charEscA=`${0o33}` 
  // No way to enter it with decimal number (27 (=0x1b)) though
const CSI=charEsc+"["
//const ANSI_CURSOR_SAVE='\0337'
const ANSI_CURSOR_SAVE=   charEsc+'7'
const ANSI_CURSOR_RESTORE=charEsc+'8'
const ANSI_CLEAR_BELOW=CSI+'J'


var ANSI_FONT_CLEAR=CSI+'0m'
var ANSI_FONT_BOLD= CSI+'1m'
var ANSI_CURSOR_UP=  function(n){return CSI+n+'A'}
//var ANSI_CURSOR_DOWN=function(n){return CSI+n+'B'}
//const ANSI_CLEAR_RIGHT=CSI+'K'
const ANSI_CURSORUP=CSI+'A'
//const ANSI_CURSORDN=CSI+'B'

const MAKESPACE_N_SAVE="\n\n"+ANSI_CURSOR_UP(2)+ANSI_CURSOR_SAVE
const MY_RESET=ANSI_CURSOR_RESTORE+ANSI_CLEAR_BELOW


// N files could be categorized as renameable after matching size and time (N OTM, N MTO, N MTM) (See list in duplicateInitial.txt)
// After looking at renamed folders (N), a further N files can be categorized as renameable. (See list in renameAdditional.txt)
//   So a final N renameables after matching size and time and folder belonging (N OTM, N MTO, N MTM) (See list in duplicateFinal.txt)


// var myRandBase63=function(n){
//   return ''.join(secrets.choice(string.ascii_lowercase +string.ascii_uppercase + string.digits+'_') for _ in range(n))
// }

// var myUUID=function(){
//   return myRandBase63(22)
// }



// var formatF_Check_Missing_In_Folder=(strPar, iStartPar, iStopPar, nSum)=>`In ${strPar} (spanning ${iStopPar-iStartPar} rows (${iStartPar}-${iStopPar-1})), ${nSum} files are missing.\n`
// var formatF_Check_Missing_Single=(iRow, strMissing, strName)=>`Row: ${iRow}, ${ANSI_FONT_BOLD}${strMissing}${ANSI_FONT_CLEAR} ${strName}\n`
// var formatF_Check_Missing_Range=(iStart, n, strMissing, strName)=>`Row: ${iStart}-${iStart+n-1} (${n}), ${ANSI_FONT_BOLD}${strMissing}${ANSI_FONT_CLEAR} ${strName}\n`

var formatF_Check_Missing_In_Folder=(strPar, iStartPar, iStopPar, nSum)=>`In ${strPar} (spanning ${iStopPar-iStartPar} rows (${iStartPar}-${iStopPar-1})), ${nSum} files are missing.\n`
var formatF_Check_Missing_Single=(iRow, strMissing, strName)=>`Row: ${iRow}, <b>${strMissing}</b> ${strName}\n`
var formatF_Check_Missing_Range=(iStart, n, strMissing, strName)=>`Row: ${iStart}-${iStart+n-1} (${n}), <b>${strMissing}</b> ${strName}\n`



class MyStatsCache{
  constructor(){ this.obj={}; }
  async stats(fiPath){
    if(!(fiPath in this.obj)) {
      var [err, fsPath]=await myRealPath(fiPath); if(err) {debugger; return [err];}
      //var [err, objStats]=await myGetStats(fiPath); 
      var [err, objStats]=await myGetStats_js(fiPath);  
      if(err) {
        if(err.code==STR_ENOENT || err.message=='noSuch') objStats={}
        else {debugger; return [err];}
      } 
      var {boDir=false, boFile=false}=objStats;
      this.obj[fiPath]={fsPath, boDir, boFile}
    }
    return [null, this.obj[fiPath]]
  }
}
gThis.myStatsCache=new MyStatsCache()

  // getHighestMissing
  // The input strFile is assumed to be a missing file.
  // Returns [err, boFile, strPar, strPath]
  //   strPath: is the highest missing
  //   strPar: is parent of strPath
  //   boFile: signifies if strPath is a file or a folder
var getHighestMissing=async function(fsDir, strFile){
  var boFile=true;
  var strPath=strFile
  while(1){
    var strPar=dirname(strPath);
    if(strPar=='') return [null, boFile, strPar, strPath]
    var [err, result]=await myStatsCache.stats(fsDir+charF+strPar); if(err) {debugger; return [err];}  
    var {fsPar, boDir}=result
    if(boDir) return [null, boFile, strPar, strPath] 
    strPath=strPar
    boFile=false
  }
}

  // For all the files in arrDb: missing files are checked if they also have missing parents.
  // Returned array ObjMissing has the same size as arrDb. ObjMissing[i] contains info about arrDb[i].
  // arrDb entries (files) that exists in fsDir has a null entry in ObjMissing.
  // arrDb entries (files) that does not exist in fsDir, has an object with info about highest missing parent.
var checkHighestMissingArr=async function(arrDb, fsDir){
  var lenDb=arrDb.length, ObjMissing=Array(lenDb).fill(null); //   [null]*lenDb
  for(var i in arrDb){
    var row=arrDb[i], strName=row.strName;
    var fsFile=fsDir+charF+strName;
    var [err, boFileFound]=await fileExist(fsFile); if(err) {debugger; return [err];}
    var obj;
    if(boFileFound) obj=null; //{"strPar":null, "strChild":null}
    else{
      var [err, boFile, strPar, strChild]= await getHighestMissing(fsDir, strName); if(err) {debugger; return [err];}
      obj={"boFile":boFile, "strPar":strPar, "strChild":strChild}
    }
    ObjMissing[i]=obj;
  }
  return [null, ObjMissing]
}



var summarizeMissing=async function(arrDb, fsDir){
  //arrDb=arrDb.slice(0,8)
  //tStop=time.time();    myConsole.printNL('checkHighestMissingArr starts, elapsed time '+str(round((tStop-tStart)*1000))+'ms')
  var [err, ObjMissing]=await checkHighestMissingArr(arrDb, fsDir); if(err) {debugger; return [err];}
  //tStop=time.time();   myConsole.printNL('checkHighestMissingArr: Done, elapsed time '+str(round((tStop-tStart)*1000))+'ms')

  // For debugging:
//   ObjMissing=[
// {'boFile': false, 'strPar': null, 'strChild': null},
// {'boFile': false, 'strPar': '', 'strChild': 'progC'},
// {'boFile': false, 'strPar': '', 'strChild': 'progC'},
// {'boFile': false, 'strPar': null, 'strChild': null},
// {'boFile': false, 'strPar': '', 'strChild': 'progD'},
// {'boFile': false, 'strPar': '', 'strChild': 'progD'},
// {'boFile': false, 'strPar': null, 'strChild': null},
// {'boFile': true, 'strPar': '', 'strChild': "a.txt"},
// {'boFile': false, 'strPar': 'a', 'strChild': 'progD'},
// {'boFile': false, 'strPar': 'a', 'strChild': 'progD'}]


  var lenDb=ObjMissing.length

    // Create objArrRangePar
  var strChildL=null, strParL=null
  var iStart=0, iStop=0, iStartPar=0, iStopPar=0, objArrRangePar={}, ArrRange=[]
  var arrRangePar
  for(var i=0;i<lenDb+1;i++){
    var objMissing=i<lenDb?ObjMissing[i]:null; 
    objMissing=objMissing??{"strPar":null, "strChild":null}
    var {strChild, strPar}=objMissing;
    if(strPar!=strParL){
      if(strChildL!=null) strChildL=myUUID(); // If parent changed then force on-going child range to end.
    }
    if(strChild!=strChildL){
      if(strChildL!=null) {iStop=i; ArrRange.push([iStart,iStop])}
      if(strChild!=null) iStart=i
    }
    if(strPar!=strParL){
      if(strParL!=null){ // parent range ended
        iStopPar=i; 
        if(!(strParL in objArrRangePar)) objArrRangePar[strParL]={"iStartPar":iStartPar, "ArrRange":[]}
        arrRangePar=objArrRangePar[strParL];
        // arrRangePar.ArrRange.push(...ArrRange); // The spread operator may cause stackoverflow in ArrRange is long.
        arrRangePar.ArrRange=arrRangePar.ArrRange.concat(ArrRange); // Same as above, but without spread operator.
        arrRangePar.iStopPar=iStopPar
        ArrRange=[]
      }
      if(strPar!=null) iStartPar=i;
    }
    strChildL=strChild
    strParL=strPar
  }

    // Calculate nSum (number of missing in each arrRangePar)
  for(var strPar in objArrRangePar){
    var arrRangePar=objArrRangePar[strPar]
    var {ArrRange}=arrRangePar;
    var nSum=0
    for(var arrRange of ArrRange){
      var [iStart, iStop]=arrRange, n=iStop-iStart
      nSum+=n
    }
    arrRangePar.nSum=nSum
  }

    // Write objArrRangePar to output
  var StrOut=[]
  for(var strPar in objArrRangePar){
    var arrRangePar=objArrRangePar[strPar]
    var {ArrRange, iStartPar, iStopPar, nSum}=arrRangePar
    if(strPar=="") strPar="(top folder)"
    StrOut.push(formatF_Check_Missing_In_Folder(strPar, iStartPar, iStopPar, nSum))
    for(var arrRange of ArrRange){
      var [iStart, iStop]=arrRange, n=iStop-iStart
      var objMissingStart=ObjMissing[iStart]
      //objMissingStop=ObjMissing[iStop-1]
      var {strChild, strPar, boFile}=objMissingStart
      if(strPar=='') strPar="(top folder)"
      if(n==1){
        var strMissing=boFile?"file":"folder";  strMissing=`Missing ${strMissing}:`;
        StrOut.push("  "+formatF_Check_Missing_Single(iStart, strMissing, strChild))
      }
      else{
        StrOut.push("  "+formatF_Check_Missing_Range(iStart, n, "Missing folder:", strChild))
      }
    }
  }
  //StrOut.push(("Summarize missing: Done"))
  var strOut=StrOut.join('')
  return [null, strOut]
}


  // Go through the db-file, for each row (file), check if the file exist, and report those in a summaristic way.
var checkSummarizeMissing=async function(arg){
  var {fiDir, charTRes, myConsole}=arg
  var [err, fsDir]=await myRealPath(fiDir); if(err) {debugger; return [err];}
  var fsDb=fsDir+charF+settings.leafDb

  setMess("Parse db...",null,true)
  var [err, strData]=await readStrFile(fsDb); if(err) return [err]
  var arrDb=parseDb(strData, charTRes);
  setMess("summarizeMissing...",null,true)
  var [err, strSum]=await summarizeMissing(arrDb, fsDir); if(err) {debugger; return [err];}
  myConsole.log(strSum)
  setMess("Summarize missing: Done")
  return [null]
}






var categorizeFile=async function(fsDir, strName){
  var fsFile=fsDir+charF+strName;
  //var boFileFound=os.path.exists(fsFile)
  var [err, boFileFound]=await fileExist(fsFile); if(err) {debugger; return [err];}
  if(boFileFound) return [null, ' ', null, null]
  else{
    var [err, boFile, strPar, flChild]= await getHighestMissing(fsDir, strName); if(err) {debugger; return [err];}
    var charMissing=boFile?'f':'d'
    return [null, charMissing, strPar, flChild]
  }
}

  // Go through the db-file, for each row (file), check if the hashcode matches the actual files hashcode  
var check=async function(arg){
  var {fiDir='.', charTRes, iStart=0, myConsole}=arg
  var [err, fsDir]=await myRealPath(fiDir); if(err) {debugger; return [err];}
  var fsDb=fsDir+charF+settings.leafDb
  var tStart=unixNow()
  var nNotFound=0, nMisMatchTimeSize=0, nMisMatchHash=0, nOK=0

  var [err, strData]=await readStrFile(fsDb); if(err) return [err]
  var arrDb=parseDb(strData, charTRes);

  var lenDb=arrDb.length
  var leafFileDbOld= basename(fsDb)



  //printNoNL(MAKESPACE_N_SAVE)
  myConsole.makeSpaceNSave()

    // Variables for "missing streaks"
  var flMissingFile="", iNotFoundFirst=0, lenFlChildL=0, charMissingL=' ', flChildL=""
  
  var nHour=0, nMin=0, nSec=0
  
  //for iRowCount, row in enumerate(arrDb):
  //for iRowCount in range(intStart,lenDb):
  for(var iRowCount=iStart;iRowCount<lenDb;iRowCount++) {
    //if(viewCheck.boCheckCancel) {setMess('Canceled'); viewCheck.boCheckCancel=false; viewCheck.butCheckCancel.hide(); return [null]}
    if(viewCheck.boCheckCancel) { return [null, "cancelled"]}
    var row=arrDb[iRowCount]
    //var {strHash:strHashOld, size:intSizeOld, mtime:intTimeOld, strName}=row;
    var {strHash:strHashOld, size:intSizeOld, mtime_ns64:intTimeOld, strName}=row;

    var fsFile=fsDir+charF+strName;   
    var [err, charMissing, strPar, flChild]=await categorizeFile(fsDir, strName); if(err) {debugger; return [err];}

    //printNoNL(MY_RESET)
    myConsole.myReset()
    
      // If streak starts
    if(charMissing=='f' && charMissingL!='f') iNotFoundFirst=iRowCount
    else if(charMissing=='d' && charMissingL!='d') iNotFoundFirst=iRowCount
    else if(charMissing=='d' && charMissingL=='d' && flChild!=flChildL) iNotFoundFirst=iRowCount

      // If streak continues
    if(charMissing=='f' && charMissingL=='f') {
      myConsole.cursorUp(); myConsole.clearBelow()
    }
    else if(charMissing=='d' && charMissingL=='d' && flChild==flChildL) {
      myConsole.cursorUp(); myConsole.clearBelow()
    }


    var [tDay,nHour,nMin,nSec,tms,tus,tns]=formatTDiff(unixNow()-tStart)
    if(charMissing=='f'){
      var nNotFoundLoc=iRowCount-iNotFoundFirst+1
      if(nNotFoundLoc==1) {
        myConsole.print(formatF_Check_Missing_Single(iNotFoundFirst, "Missing file:", strName))
        myConsole.makeSpaceNSave()
      }
      else{
        //var strTmp=os.path.dirname(strName)
        var strTmp=dirname(strName);
        if(strTmp=="") strTmp="(top folder)"
        myConsole.print(formatF_Check_Missing_Range(iNotFoundFirst, nNotFoundLoc, "Missing files in:", strTmp))
        myConsole.makeSpaceNSave()
      }
    } else if(charMissing=='d'){
      nNotFoundLoc=iRowCount-iNotFoundFirst+1
      if(nNotFoundLoc==1){
        myConsole.print(formatF_Check_Missing_Single(iNotFoundFirst, "Missing folder:", flChild))
        myConsole.makeSpaceNSave()
      }
      else{
        myConsole.print(formatF_Check_Missing_Range(iNotFoundFirst, nNotFoundLoc, "Missing folder:", flChild))
        myConsole.makeSpaceNSave()
      }
    } else{
      var strTmp=`${nHour.myPad0(2)}:${nMin.myPad0(2)}:${nSec.myPad0(2)}`
      myConsole.printNL(`${strTmp}, Checking row: ${iRowCount} ${strName}`)
    }
    charMissingL=charMissing; flChildL=flChild
    if(charMissing!=' ') {nNotFound+=1; continue}


      // Calculate hash
    var [err, strHash]=await myMD5(row, fsDir, false); //.decode("utf-8")   
    if(strHash!=strHashOld){ // If hashes mismatches
        // Check modTime and size (perhaps the user forgott to run sync before running check
      var [err, objT]=await myGetStats(fsFile); if(err) {debugger; return [err];}
      //var {size:intSizeNew, mtime:intTimeNew}=objT;
      var {size:intSizeNew, mtime_ns64:intTimeNew}=objT;
      var boTMatch=intTimeNew==intTimeOld,    boSizeMatch=intSizeNew==intSizeOld
      myConsole.myReset()
      var StrTmp=[]
      if(!boTMatch || !boSizeMatch ){// If meta data mismatches
        var strBase=basename(strName)
        if(strBase==leafFileDbOld){
          //var strTmp=`${MY_RESET}Row: ${iRowCount}, (${strName} is ignored)`;      StrTmp.push(strTmp)
          var strTmp=`Row: ${iRowCount}, (${strName} is ignored)`;      StrTmp.push(strTmp)
        }
        else{
          //var strTmp = `${MY_RESET}Row: ${iRowCount}`;      StrTmp.push(strTmp)
          var strTmp = `Row: ${iRowCount}`;      StrTmp.push(strTmp)

          var StrLab=[];
          if(!boSizeMatch) StrLab.push("SIZE")
          if(!boTMatch) StrLab.push("MTIME")
          var strLab=StrLab.join(' and ');
          StrTmp.push(`<b>${strLab} MISMATCH!!!</b> (remember to <b>SYNC</b> the database before checking)`)

          if(!boSizeMatch) {
            var strTmp=`<b>size</b> (db/tree): ${intSizeOld}/${intSizeNew}`;    StrTmp.push(strTmp)
          }
          if(!boTMatch) {
            var tDiff64=intTimeNew-intTimeOld;
            var tDiffS=Number(tDiff64/BigInt(1e9));
            if(tDiffS==0){var tDiffHuman=Number(tDiff64%BigInt(1e9)), charUnit='ns' }
            else { var [tDiffHuman, charUnit]=getSuitableTimeUnit(tDiffS) }
            var strTmp=`<b>tDiff</b> (tree-db): ${tDiffHuman}${charUnit}`;    StrTmp.push(strTmp)
          }
          
          var strTmpB=strName;    StrTmp.push(strTmpB)
        }
        
        nMisMatchTimeSize+=1
      }
      else{// Meta data matches
        //var strTmp = (`${MY_RESET}Row: ${iRowCount}, `+"<b>Hash Mismatch</b>");    StrTmp.push(strTmp)
        var strTmp = (`Row: ${iRowCount}, <b>Hash Mismatch</b>`);    StrTmp.push(strTmp)
        StrTmp.push(`(db/tree): ${strHashOld} / ${strHash}`);   
        StrTmp.push(strName);    
        nMisMatchHash+=1
      }
      
      var strTmp=StrTmp.join(", ")+"\n"
      myConsole.print(strTmp); myConsole.makeSpaceNSave()
    }
    
    else nOK+=1;  // Hashes match
  }
  var [tDay,nHour,nMin,nSec,tms,tus,tns]=formatTDiff(unixNow()-tStart)
  var StrSum=[]
  StrSum.push(`RowCount: ${iRowCount}`)
  var strTmp=`NotFound: ${nNotFound}`;   if(nNotFound) strTmp=`<b>${strTmp}</b>`;      StrSum.push(strTmp)
  var strTmp=`MisMatchTimeSize: ${nMisMatchTimeSize}`;   if(nMisMatchTimeSize) strTmp=`<b>${strTmp}</b>`;      StrSum.push(strTmp)
  var strTmp=`MisMatchHash: ${nMisMatchHash}`;   if(nMisMatchHash) strTmp=`<b>${strTmp}</b>`;     StrSum.push(strTmp)
  StrSum.push(`OK: ${nOK}`)
  var strSum=StrSum.join(', ')

  var strTmp=`${nHour.myPad0(2)}:${nMin.myPad0(2)}:${nSec.myPad0(2)}`
  myConsole.myReset()
  myConsole.printNL(`Time: ${strTmp}, Done (${strSum})`); //resetMess()
  if(nMisMatchTimeSize) myConsole.printNL(`<b>META (SIZE/MTIME) MISMATCHES!!!</b> (on ${nMisMatchTimeSize} file${pluralS(nMisMatchTimeSize)}) This happens when the db wasn't <b>SYNCED</b> before checking.`)

  return [null, "Check finished"]
}

File: buvtPyScript/libCheck.py

from lib import *
import time
import stat
import uuid

import globvar
from lib import *
from libFs import *
from libParse import *



ANSI_FONT_CLEAR=""
ANSI_FONT_BOLD=""
def formatF_Check_Missing_In_Folder(strPar, iStartPar, iStopPar, nSum):
  nSpan=iStopPar-iStartPar
  charS=pluralS(nSpan)
  #strFileIs='file is' if(nSum==1) else 'files are'
  strFileIs=pluralS(nSum, 'file is', 'files are')
  return f"In {strPar} (spanning {nSpan} row{charS} ({iStartPar}-{iStopPar-1})), {nSum} {strFileIs} missing.\n"
def formatF_Check_Missing_Single(iRow, strMissing, strName):
  return f"Row: {iRow}, {strMissing.upper()}: {strName}\n"
def formatF_Check_Missing_Range(iStart, n, strMissing, strName):
  return f"Row: {iStart}-{iStart+n-1} ({n}), {strMissing.upper()}: {strName}\n"



# N files could be categorized as renameable after matching size and time (N OTM, N MTO, N MTM) (See list in duplicateInitial.txt)
# After looking at renamed folders (N), a further N files can be categorized as renameable. (See list in renameAdditional.txt)
#   So a final N renameables after matching size and time and folder belonging (N OTM, N MTO, N MTM) (See list in duplicateFinal.txt)


# FMT_Check_Missing_In_Folder=   "In %s (spanning %d rows (%d-%d)), %d file(s) are missing.\n"
# FMT_Check_Missing_Single=    "Row: %d, " +ANSI_FONT_BOLD+"%s"+ANSI_FONT_CLEAR+" %s\n"
# FMT_Check_Missing_Range=   "Row: %d-%d (%d), " +ANSI_FONT_BOLD+"%s"+ANSI_FONT_CLEAR+" %s\n"





class MyRealPath:
  def __init__(self):
    self.obj={}
  def realPath(self, fsPath):
    if(fsPath not in self.obj): 
      self.obj[fsPath]=myRealPathf(fsPath)
    return self.obj[fsPath]
myRealPath=MyRealPath()

  # getHighestMissing
  # The input strFile is assumed to be a missing file.
  # Returns boFile, strPar, strPath
  #   strPath: is the highest missing
  #   strPar: is parent of strPath
  #   boFile: signifies if strPath is a file or a folder
def getHighestMissing(fsDir, strFile): 
  boFile=True
  strPath=strFile
  while(1):
    strPar=os.path.dirname(strPath)
    if(strPar==''): return boFile, strPar, strPath
    fsPar=myRealPath.realPath(fsDir+charF+strPar)
    if( os.path.isdir(fsPar)): return boFile, strPar, strPath 
    strPath=strPar
    boFile=False


def categorizeFile(fsDir, strName):
  fsFile=fsDir+charF+strName;     boFileFound=os.path.lexists(fsFile) # lexists: doesn't follow links (returns true on broken links)
  if(boFileFound): return ' ', None, None
  else:
    boFile, strPar, flChild= getHighestMissing(fsDir, strName)
    charMissing='f' if(boFile) else 'd'
    return charMissing, strPar, flChild


  # Go through the db-file, for each row (file), check if the hashcode matches the actual files hashcode  
def check(**args):
  fiDbDir=myRealPathf(args["fiDbDir"])
  fiDb=args["fiDb"]
  fsDb=myRealPathf(fiDb) 
  iStart=int(args["iStart"])
  charTRes=args["charTRes"]

  tStart=time.time()
  myConsole=globvar.myConsole
  nNotFound=0; nMisMatchTimeSize=0; nMisMatchHash=0; nOK=0

  #myConsole.print(MAKESPACE_N_SAVE)
  myConsole.makeSpaceNSave()
  
  myConsole.print('parsing db...')
  err, arrDb=parseDbWType(fsDb, charTRes)
  if(err): print(err["strTrace"], file=sys.stderr); return
  #if(err): myConsole.error(err["strTrace"]); return

  lenDb=len(arrDb)
  leafFileDbOld= os.path.basename(fsDb)


    # Variables for "missing streaks"
  flMissingFile=""
  iNotFoundFirst=0
  lenFlChildL=0
  charMissingL=' '; flChildL=""
  
  nHour=0; nMin=0; nSec=0
  
  #for iRowCount, row in enumerate(arrDb):
  for iRowCount in range(iStart,lenDb):
    row=arrDb[iRowCount]
    strHashOld=row["strHash"]; intSizeOld=row["size"]; intTimeOld=row["mtime_ns64Floored"]; strName=row["strName"]

    fsFile=fiDbDir+charF+strName;   
    charMissing, strPar, flChild=categorizeFile(fiDbDir, strName)

    #myConsole.print(MY_RESET)
    myConsole.myReset()
    
      # If streak starts
    if(charMissing=='f' and charMissingL!='f'): iNotFoundFirst=iRowCount
    elif(charMissing=='d' and charMissingL!='d'): iNotFoundFirst=iRowCount
    elif(charMissing=='d' and charMissingL=='d' and flChild!=flChildL): iNotFoundFirst=iRowCount

      # If streak continues
    if(charMissing=='f' and charMissingL=='f'):
      myConsole.cursorUp(); myConsole.clearBelow()
      #myConsole.print(ANSI_CURSORUP+ANSI_CLEAR_BELOW)
    elif(charMissing=='d' and charMissingL=='d' and flChild==flChildL):
      myConsole.cursorUp(); myConsole.clearBelow()
      #myConsole.print(ANSI_CURSORUP+ANSI_CLEAR_BELOW)


    tDay,nHour,nMin,nSec,tms,tus,tns=formatTDiff(time.time()-tStart)
    if(charMissing=='f'):
      nNotFoundLoc=iRowCount-iNotFoundFirst+1
      if(nNotFoundLoc==1):
        #myConsole.print((FMT_Check_Missing_Single+MAKESPACE_N_SAVE) %(iNotFoundFirst, "Missing file:", strName))
        myConsole.print(formatF_Check_Missing_Single(iNotFoundFirst, "Missing file", strName))
        myConsole.makeSpaceNSave()
      else:
        strTmp=os.path.dirname(strName)
        if(strTmp==""): strTmp="(top folder)"
        #myConsole.print((FMT_Check_Missing_Range+MAKESPACE_N_SAVE) %(iNotFoundFirst, iRowCount, nNotFoundLoc, "Missing files in:", strTmp))
        myConsole.print(formatF_Check_Missing_Range(iNotFoundFirst, nNotFoundLoc, "Missing files in", strTmp))
        myConsole.makeSpaceNSave()
    elif(charMissing=='d'):
      nNotFoundLoc=iRowCount-iNotFoundFirst+1
      if(nNotFoundLoc==1):
        #myConsole.print((FMT_Check_Missing_Single+MAKESPACE_N_SAVE) %(iNotFoundFirst, "Missing folder:", flChild))
        myConsole.print(formatF_Check_Missing_Single(iNotFoundFirst, "Missing folder", flChild))
        myConsole.makeSpaceNSave()
      else:
        #myConsole.print((FMT_Check_Missing_Range+MAKESPACE_N_SAVE) %(iNotFoundFirst, iRowCount, nNotFoundLoc, "Missing folder:", flChild))
        myConsole.print(formatF_Check_Missing_Range(iNotFoundFirst, nNotFoundLoc, "Missing folder", flChild))
        myConsole.makeSpaceNSave()
    else: 
      #myConsole.print(("%d:%02d:%02d, "+ANSI_FONT_BOLD+"Checking row:"+ANSI_FONT_CLEAR+" %d %s\n") %(nHour, nMin, nSec, iRowCount, strName))
      myConsole.printNL(("%d:%02d:%02d, Checking row: %d %s") %(nHour, nMin, nSec, iRowCount, strName))
    
    charMissingL=charMissing; flChildL=flChild
    if(charMissing!=' '): nNotFound+=1; continue


      # Calculate hash
    #strHash=myMD5(fsFile).decode("utf-8")
    strHash=myMD5W(row, fiDbDir)
    if(strHash!=strHashOld):  # If hashes mismatches
        # Check modTime and size (perhaps the user forgott to run sync before running check
      st = os.lstat(fsFile); st_size=st.st_size; st_mtime=st.st_mtime; st_mtime_ns=st.st_mtime_ns
      #intTimeNew=math.floor(st_mtime)
      intTimeNew=st_mtime_ns
      boTMatch=intTimeNew==intTimeOld;    boSizeMatch=st_size==intSizeOld
      myConsole.myReset()
      StrTmp=[]
      if(not boTMatch or not boSizeMatch ): # If meta data mismatches
        strBase=os.path.basename(strName)
        if(strBase==leafFileDbOld):
          #strTmp=(MY_RESET+"Row: %d, (%s is ignored)") %(iRowCount, strName);   StrTmp.append(strTmp)
          strTmp=("Row: %d, (%s is ignored)") %(iRowCount, strName);   StrTmp.append(strTmp)
        else:
          #strTmp = (MY_RESET+"Row: %d") %(iRowCount);   StrTmp.append(strTmp)
          strTmp = ("Row: %d") %(iRowCount);   StrTmp.append(strTmp)
          

          StrLab=[]
          if(not boSizeMatch): StrLab.append("SIZE")
          if(not boTMatch): StrLab.append("MTIME")
          strLab=' and '.join(StrLab)
          #strLabB=ANSI_FONT_BOLD+strLab+" MISMATCH!!!"+ANSI_FONT_CLEAR+" (run SYNC before checking)"
          strLabB="(Run SYNC before checking)"
          #StrTmp.append(strLabB)
          if(not boSizeMatch): strTmp = (ANSI_FONT_BOLD+"size"+ANSI_FONT_CLEAR+" (db/file): %d/%d") %(intSizeOld, st_size);    StrTmp.append(strTmp)
          if(not boTMatch): 
            tDiffNs=intTimeNew-intTimeOld; tDiff=tDiffNs/1e9; tDiffHuman, charUnit=getSuitableTimeUnit(tDiff)
            #strTmp = (ANSI_FONT_BOLD+"tDiff"+ANSI_FONT_CLEAR+" (file-db): %i%c") %(tDiffHuman, charUnit);
            strTmp='behind' if(tDiffNs>0) else 'ahead'
            strTmp = f'mtime: db is {str(math.floor(abs(tDiffHuman)))}{charUnit} {strTmp}';
            StrTmp.append(strTmp)
          
          StrTmp.append(strLabB)
          strTmpB = "%s" %(strName);    StrTmp.append(strTmpB)
        
        nMisMatchTimeSize+=1
      else: # Meta data matches
        #strTmp = (MY_RESET+"Row: %d, "+ANSI_FONT_BOLD+"Hash Mismatch"+ANSI_FONT_CLEAR) %(iRowCount);    StrTmp.append(strTmp)
        strTmp = ("Row: %d, Hash Mismatch!!!") %(iRowCount);    StrTmp.append(strTmp)
        StrTmp.append("(db/file): "+strHashOld+" / "+strHash);   
        StrTmp.append(strName);    
        nMisMatchHash+=1
      
      strTmp=", ".join(StrTmp)
      myConsole.printNL(strTmp)
      #myConsole.print(MAKESPACE_N_SAVE)
      myConsole.makeSpaceNSave()
    
    else: nOK+=1;  # Hashes match

  tDay,nHour,nMin,nSec,tms,tus,tns=formatTDiff(time.time()-tStart)
  StrSum=[]
  StrSum.append("nRowTested/nRowTot: %d/%d")
  strTmp="NotFound: %d";
  if(nNotFound): strTmp=ANSI_FONT_BOLD+strTmp+ANSI_FONT_CLEAR
  StrSum.append(strTmp)
  strTmp="MisMatchTimeSize: %d";
  if(nMisMatchTimeSize): strTmp=ANSI_FONT_BOLD+strTmp+ANSI_FONT_CLEAR
  StrSum.append(strTmp)
  strTmp="MisMatchHash: %d";
  if(nMisMatchHash): strTmp=ANSI_FONT_BOLD+strTmp+ANSI_FONT_CLEAR
  StrSum.append(strTmp)
  StrSum.append("OK: %d")
  strSum=', '.join(StrSum)

  myConsole.myReset()
  myConsole.printNL(("Time: %d:%02d:%02d, Done ("+strSum+")") %(nHour,nMin,nSec, lenDb-iStart, lenDb, nNotFound, nMisMatchTimeSize, nMisMatchHash, nOK))
  if(nMisMatchTimeSize):
    #strLab=ANSI_FONT_BOLD+"META (SIZE/MTIME) MISMATCHES!!!"+ANSI_FONT_CLEAR+" (on "+str(nMisMatchTimeSize)+" files) This happens when the db-file hasn't been SYNC-ed before checking"
    strLab=str(nMisMatchTimeSize)+" cases where meta-data (size/mtime) mismatches. This happens when the db hasn't been SYNC-ed before checking."
    myConsole.printNL(strLab)


  return ""


